File: backend/app/services/weather.py
import logging
from datetime import UTC, datetime
from typing import Any

import httpx
from fastapi import HTTPException

from app.cache.manager import CacheManager
from app.config import Settings
from app.models.weather import MetarObservation, MetarResponse, SigmetFeature, SigmetResponse

logger = logging.getLogger(__name__)


class WeatherService:
    def __init__(self, settings: Settings, client: httpx.AsyncClient, cache: CacheManager) -> None:
        self._settings = settings
        self._client = client
        self._cache = cache

    async def get_sigmets(self) -> SigmetResponse:
        return await self._cache.get_or_set(
            "weather:sigmets",
            self._settings.weather_cache_ttl_seconds,
            self._fetch_sigmets,
        )

    async def get_metars(self, ids: list[str]) -> MetarResponse:
        station_ids = _normalize_station_ids(ids)
        key = f"weather:metars:{','.join(station_ids)}"
        return await self._cache.get_or_set(
            key,
            self._settings.weather_cache_ttl_seconds,
            lambda: self._fetch_metars(station_ids),
        )

    async def _fetch_sigmets(self) -> SigmetResponse:
        response = await self._client.get(
            self._settings.weather_sigmet_url,
            params={"format": "geojson"},
            headers={"User-Agent": self._settings.user_agent, "Accept": "application/geo+json, application/json"},
        )
        if response.status_code == 204:
            geojson = {"type": "FeatureCollection", "features": []}
            return SigmetResponse(count=0, geojson=geojson, sigmets=[])
        if response.status_code == 429:
            raise HTTPException(429, "AviationWeather rate limit reached; retry shortly.")
        if response.status_code >= 400:
            logger.warning("AviationWeather SIGMET request failed: %s", response.status_code)
            raise HTTPException(502, "AviationWeather SIGMET request failed.")

        geojson = response.json()
        features = geojson.get("features") or []
        sigmets = [_parse_sigmet(feature) for feature in features]
        return SigmetResponse(count=len(sigmets), geojson=geojson, sigmets=sigmets)

    async def _fetch_metars(self, ids: list[str]) -> MetarResponse:
        response = await self._client.get(
            self._settings.weather_metar_url,
            params={"ids": ",".join(ids), "format": "json"},
            headers={"User-Agent": self._settings.user_agent, "Accept": "application/json"},
        )
        if response.status_code == 204:
            return MetarResponse(count=0, observations=[])
        if response.status_code == 429:
            raise HTTPException(429, "AviationWeather rate limit reached; retry shortly.")
        if response.status_code >= 400:
            logger.warning("AviationWeather METAR request failed: %s", response.status_code)
            raise HTTPException(502, "AviationWeather METAR request failed.")

        payload = response.json()
        rows = payload if isinstance(payload, list) else payload.get("data", [])
        observations = [_parse_metar(row) for row in rows if isinstance(row, dict)]
        return MetarResponse(count=len(observations), observations=observations)


def _parse_sigmet(feature: dict[str, Any]) -> SigmetFeature:
    properties = feature.get("properties") or {}
    sigmet_type = (
        properties.get("hazard")
        or properties.get("hazardType")
        or properties.get("type")
        or properties.get("airSigmetType")
    )
    raw_text = properties.get("rawText") or properties.get("raw_text") or properties.get("text")
    severity = _infer_severity(sigmet_type, raw_text or "")
    return SigmetFeature(
        id=properties.get("id") or properties.get("airSigmetId") or feature.get("id"),
        type=sigmet_type,
        severity=severity,
        raw_text=raw_text,
        geometry=feature.get("geometry"),
        properties=properties,
    )


def _parse_metar(row: dict[str, Any]) -> MetarObservation:
    return MetarObservation(
        station=row.get("icaoId") or row.get("station_id") or row.get("id") or "UNKNOWN",
        observed=_observed_time(row.get("obsTime") or row.get("observation_time") or row.get("reportTime")),
        raw_text=row.get("rawOb") or row.get("raw_text") or row.get("raw"),
        flight_category=row.get("fltCat") or row.get("flight_category"),
        latitude=_float_or_none(row.get("lat") or row.get("latitude")),
        longitude=_float_or_none(row.get("lon") or row.get("longitude")),
        temperature_c=_float_or_none(row.get("temp") or row.get("temp_c")),
        dewpoint_c=_float_or_none(row.get("dewp") or row.get("dewpoint_c")),
        wind_direction_deg=_int_or_none(row.get("wdir") or row.get("wind_dir_degrees")),
        wind_speed_kt=_int_or_none(row.get("wspd") or row.get("wind_speed_kt")),
        wind_gust_kt=_int_or_none(row.get("wgst") or row.get("wind_gust_kt")),
        visibility_sm=_float_or_none(row.get("visib") or row.get("visibility_statute_mi")),
        altimeter_in_hg=_altimeter_in_hg(row.get("altim") or row.get("altim_in_hg")),
        clouds=row.get("clouds") if isinstance(row.get("clouds"), list) else [],
    )


def _normalize_station_ids(ids: list[str]) -> list[str]:
    station_ids = []
    for station_id in ids:
        cleaned = station_id.strip().upper()
        if cleaned and cleaned.isalnum() and 3 <= len(cleaned) <= 4:
            station_ids.append(cleaned)
    unique = list(dict.fromkeys(station_ids))
    if not unique:
        unique = ["VIDP", "VABB", "VOBL", "VOMM", "VOHS", "VECC", "VAAH"]
    return unique[:12]


def _float_or_none(value: Any) -> float | None:
    if isinstance(value, str) and value.endswith("+"):
        value = value[:-1]
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def _int_or_none(value: Any) -> int | None:
    try:
        return int(float(value))
    except (TypeError, ValueError):
        return None


def _observed_time(value: Any) -> str | None:
    if value is None:
        return None
    if isinstance(value, int | float):
        return datetime.fromtimestamp(value, tz=UTC).isoformat()
    return str(value)


def _altimeter_in_hg(value: Any) -> float | None:
    altimeter = _float_or_none(value)
    if altimeter is None:
        return None
    if altimeter > 100:
        return round(altimeter * 0.0295299830714, 2)
    return altimeter


def _infer_severity(sigmet_type: str | None, raw_text: str) -> str:
    value = f"{sigmet_type or ''} {raw_text}".upper()
    if "EXTREME" in value:
        return "extreme"
    if "SEV" in value or "CONV" in value or "VOLCANIC" in value or "ASH" in value:
        return "severe"
    if "MOD" in value:
        return "moderate"
    return "advisory"

File: backend-node/src/services/weather.js
// ── services/weather.js ───────────────────────────────────────────────────────
// WeatherService — port of Python backend/app/services/weather.py
// Fetches SIGMETs and METARs from aviationweather.gov and caches them.
// ──────────────────────────────────────────────────────────────────────────────

import { httpErr } from '../utils/bbox.js';

export class WeatherService {
  /**
   * @param {import('../config.js').default} settings
   * @param {import('axios').AxiosInstance}  client
   * @param {import('../cache/manager.js').CacheManager} cache
   */
  constructor(settings, client, cache) {
    this._settings = settings;
    this._client   = client;
    this._cache    = cache;
  }

  // ── Public API ──────────────────────────────────────────────────────────────

  async getSigmets() {
    return this._cache.getOrSet(
      'weather:sigmets',
      this._settings.weatherCacheTtlSeconds,
      () => this._fetchSigmets(),
    );
  }

  async getMetars(ids) {
    const stationIds = normalizeStationIds(ids);
    const key = `weather:metars:${stationIds.join(',')}`;
    return this._cache.getOrSet(
      key,
      this._settings.weatherCacheTtlSeconds,
      () => this._fetchMetars(stationIds),
    );
  }

  // ── Private ─────────────────────────────────────────────────────────────────

  async _fetchSigmets() {
    const urls = [
      this._settings.weatherSigmetUrl,
      this._settings.weatherIsigmetUrl,
    ];

    const requests = urls.map(url =>
      this._client.get(url, {
        params: { format: 'geojson' },
        headers: {
          'User-Agent': this._settings.userAgent,
          Accept: 'application/geo+json, application/json',
        },
      }).catch(err => {
        if (err.response) {
          const status = err.response.status;
          if (status === 429) {
            throw httpErr(429, 'AviationWeather rate limit reached; retry shortly.');
          }
          console.warn(`[WeatherService] SIGMET request failed for ${url}: ${status}`);
          return { status, data: null };
        }
        console.warn(`[WeatherService] SIGMET request timed out/failed for ${url}: ${err.message}`);
        return { status: 504, data: null };
      })
    );

    const responses = await Promise.all(requests);

    // Merge features from successful responses
    let mergedFeatures = [];
    for (const res of responses) {
      if (res && res.data && Array.isArray(res.data.features)) {
        mergedFeatures.push(...res.data.features);
      }
    }

    const parsedSigmets = mergedFeatures.map(parseSigmet);
    const geojson = {
      type: 'FeatureCollection',
      features: mergedFeatures,
    };

    return {
      source: 'aviationweather.gov (domestic + international)',
      count: parsedSigmets.length,
      geojson,
      sigmets: parsedSigmets,
    };
  }

  async _fetchMetars(ids) {
    let response;
    try {
      response = await this._client.get(this._settings.weatherMetarUrl, {
        params: { ids: ids.join(','), format: 'json' },
        headers: {
          'User-Agent': this._settings.userAgent,
          Accept: 'application/json',
        },
      });
    } catch (err) {
      if (err.response) {
        const status = err.response.status;
        if (status === 429) throw httpErr(429, 'AviationWeather rate limit reached; retry shortly.');
        console.warn(`[WeatherService] METAR request failed: ${status}`);
        throw httpErr(502, 'AviationWeather METAR request failed.');
      }
      throw httpErr(504, 'AviationWeather timed out or is unreachable.');
    }

    if (response.status === 204) return { source: 'aviationweather.gov', count: 0, observations: [] };

    const payload = response.data;
    const rows = Array.isArray(payload) ? payload : (payload.data ?? []);
    const observations = rows.filter(r => r && typeof r === 'object').map(parseMetar);
    return { source: 'aviationweather.gov', count: observations.length, observations };
  }
}

// ── Pure helpers ──────────────────────────────────────────────────────────────

function parseSigmet(feature) {
  const properties = feature.properties ?? {};
  const sigmetType =
    properties.hazard ??
    properties.hazardType ??
    properties.type ??
    properties.airSigmetType ??
    null;
  const rawText =
    properties.rawAirSigmet ??
    properties.rawText ??
    properties.raw_text ??
    properties.text ??
    null;
  const severity = inferSeverity(sigmetType, rawText ?? '');
  return {
    id:         properties.id ?? properties.airSigmetId ?? feature.id ?? null,
    type:       sigmetType,
    severity,
    raw_text:   rawText,
    geometry:   feature.geometry ?? null,
    properties,
  };
}

function parseMetar(row) {
  return {
    station:             row.icaoId ?? row.station_id ?? row.id ?? 'UNKNOWN',
    observed:            observedTime(row.obsTime ?? row.observation_time ?? row.reportTime),
    raw_text:            row.rawOb ?? row.raw_text ?? row.raw ?? null,
    flight_category:     row.fltCat ?? row.flight_category ?? null,
    latitude:            floatOrNull(row.lat ?? row.latitude),
    longitude:           floatOrNull(row.lon ?? row.longitude),
    temperature_c:       floatOrNull(row.temp ?? row.temp_c),
    dewpoint_c:          floatOrNull(row.dewp ?? row.dewpoint_c),
    wind_direction_deg:  intOrNull(row.wdir ?? row.wind_dir_degrees),
    wind_speed_kt:       intOrNull(row.wspd ?? row.wind_speed_kt),
    wind_gust_kt:        intOrNull(row.wgst ?? row.wind_gust_kt),
    visibility_sm:       floatOrNull(row.visib ?? row.visibility_statute_mi),
    altimeter_in_hg:     altimeterInHg(row.altim ?? row.altim_in_hg),
    clouds:              Array.isArray(row.clouds) ? row.clouds : [],
  };
}

function normalizeStationIds(ids) {
  const unique = [];
  const seen = new Set();
  for (const id of ids) {
    const cleaned = id.trim().toUpperCase();
    if (cleaned && /^[A-Z0-9]{3,4}$/.test(cleaned) && !seen.has(cleaned)) {
      unique.push(cleaned);
      seen.add(cleaned);
    }
  }
  if (unique.length === 0) return ['KJFK', 'KLAX', 'KORD', 'KATL', 'KDFW', 'KDEN', 'KSFO', 'KMIA'];
  return unique.slice(0, 12);
}

function observedTime(value) {
  if (value == null) return null;
  if (typeof value === 'number') return new Date(value * 1000).toISOString();
  return String(value);
}

function inferSeverity(sigmetType, rawText) {
  const val = `${sigmetType ?? ''} ${rawText}`.toUpperCase();
  if (val.includes('EXTREME')) return 'extreme';
  if (val.includes('SEV') || val.includes('CONV') || val.includes('VOLCANIC') || val.includes('ASH'))
    return 'severe';
  if (val.includes('MOD')) return 'moderate';
  return 'advisory';
}

function floatOrNull(value) {
  if (typeof value === 'string' && value.endsWith('+')) value = value.slice(0, -1);
  const n = parseFloat(value);
  return isNaN(n) ? null : n;
}

function intOrNull(value) {
  const n = parseInt(value, 10);
  return isNaN(n) ? null : n;
}

function altimeterInHg(value) {
  const alt = floatOrNull(value);
  if (alt === null) return null;
  if (alt > 100) return Math.round(alt * 0.0295299830714 * 100) / 100;
  return alt;
}
